modules/aws/cloudfront-s3/main.tf

```diff
@@ -0,0 +1,215 @@
+  bucket = var.cf_origin_bucket_name
+}
+
+resource "aws_s3_bucket_policy" "bucket_policy" {
+  bucket = data.aws_s3_bucket.cf_origin_bucket.id
+  policy = data.aws_iam_policy_document.s3_policy.json
+}
+
+##Module for creating logging bucket.
+##Bucket allows to store logs files generated by cloudfront
+data "aws_canonical_user_id" "current" {}
+data "aws_caller_identity" "current_account" {}
+
+module "cf-logging-bucket" {
+  create_bucket = var.logging ? true : false ## Bucket would be created only when logging is enabled
+  source = "terraform-aws-modules/s3-bucket/aws"
+  version = "3.3.0"
+  bucket = var.logging ? "cf-accesslog-storage-${random_pet.this[0].id}" : null
+
+  # Allow deletion of non-empty bucket
+  force_destroy = true
+  acl    = null
+  ##ACL for Bucket. 
+  ##1. Full Control to awslogsdelivery and to the user who is creating bucket
+  grant = [{
+    type        = "CanonicalUser"
+    permission = "FULL_CONTROL"
+    id          = data.aws_canonical_user_id.current.id
+    }, {
+    type        = "CanonicalUser"
+    permission = "FULL_CONTROL"
+    id          = "c4c1ede66af53448b93c283ce9448c4ba468c9432aa01d700d3878632f77d2d0"
+    # Ref. https://github.com/terraform-providers/terraform-provider-aws/issues/12512
+    # Ref. https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/AccessLogs.html
+  }]
+  server_side_encryption_configuration = var.logging ? {
+    rule = {
+      apply_server_side_encryption_by_default = {
+        kms_master_key_id = aws_kms_key.objects[0].arn
+        sse_algorithm     = "aws:kms"
+      }
+    }
+  }: {}
+} 
+
+##Module for creating KMS key to encrpt the logging bucket
+resource "aws_kms_key" "objects" {
+  count = var.logging ? 1 : 0
+  description             = "KMS key is used to encrypt bucket objects"
+  deletion_window_in_days = 7
+  policy = data.aws_iam_policy_document.kms_policy_document[0].json
+}
+
+#Since we are creating an KMS key to encrypt and decrypt objects in logging bucket we need to add following permissions to KMS Key
+# Ref: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/AccessLogs.html#AccessLogsBucketAndFileOwnership ( Required key policy for SSE-KMS buckets )
+data "aws_iam_policy_document" "kms_policy_document" {
+  count = var.logging ? 1 : 0
+  statement {
+    sid = "Allow CloudFront to use the key to deliver logs"
+
+    actions = [
+      "kms:GenerateDataKey*",
+    ]
+    principals {
+      type        = "Service"
+      identifiers = ["delivery.logs.amazonaws.com"]
+    }
+    resources = [
+      "*"
+    ]
+  }
+  statement {
+    sid = "Enable IAM User Permissions"
+
+    actions = [
+      "kms:*",
+    ]
+    principals {
+      type        = "AWS"
+      identifiers = ["arn:aws:iam::${data.aws_caller_identity.current_account.account_id}:root"]
+    }
+    resources = [
+      "*"
+    ]
+  }
+}
+
+resource "random_pet" "this" {
+  count = var.logging ? 1 : 0
+  length = 2
+}
```

Comment:
???

Reply:
Ok, We are creating a new bucket when logging is enabled, as we know bucket names should be unique, what i am doing is to make bucket names unique i am just referencing above resource in bucketnames.
"cf-accesslog-storage-${random_pet.this[0].id}"